### PR TITLE
[Help] Help Fixes and Enhancements

### LINKED ISSUE #365: [Help] Extract command descriptions from method docstring
This is a known gap we should fill in. Lost of methods have pretty good summary in its docstring which can be used for help. And cli is already loading the param section in docstring for help, and we should make it complete 
The effort involved would be to parse the doc-string in finer grain to extract out the summary

### LINKED ISSUE #447: [Help] help doesn't support short descriptions with line breaks
Currently no commands use this, but if we use it when more help content rolls in, we'll need to implement this first

### LINKED ISSUE #680: [Register command] Parameter comments with ":" are not parsed
Instead the comment is blank.  This frequently happens with text like "Example: ..."

### PR DESCRIPTION
**Pull request by tjprescott**
This PR fixes #365, fixes #447 and fixes #680. The first sentence of the docstring will be parsed as the short description and any subsequent sentences will be parsed as the long description. This largely takes care of issue #707 (although it doesn't address how adequate the reflected help is). Of course, you can override any of this through the YAML help file.

Additionally, this PR fixes an issue where choice lists would appear in random order in help.

### CODE REVIEW

**Comment by tjprescott** [comment]
As an example for `az vm -h`

```
Group
    az vm

Subgroups:
    ...

Commands:
    capture               : Captures the VM by copying virtual hard disks of the VM and outputs a
                            template that can be used to create similar VMs.
    create                : Create an Azure Virtual Machine.
    deallocate            : Shuts down the Virtual Machine and releases the compute resources.
    delete                : The operation to delete a virtual machine.
    generalize            : Sets the state of the VM as Generalized.
    get-instance-view
    list                  : List Virtual Machines.
    list-ip-addresses     : Get IP addresses from one or more Virtual Machines.
    list-sizes            : Lists all available virtual machine sizes for a subscription in a
                            location.
    list-usage            : Lists compute usages for a subscription.
    list-vm-resize-options: Lists all available virtual machine sizes it can be resized to for a
                            virtual machine.
    open-port             : Opens a VM to all inbound traffic and protocols by adding a security
                            rule to the network security group (NSG) that is attached to the VM's
                            network interface (NIC) or subnet.
    redeploy              : The operation to redeploy a virtual machine.
    resize                : Update vm size.
    restart               : The operation to restart a virtual machine.
    show                  : The operation to get a virtual machine.
    start                 : The operation to start a virtual machine.
    stop                  : The operation to power off (stop) a virtual machine.
    update                : Update VM properties.
```

File: src/azure/cli/commands/_introspection.py [hunk 1]
```diff
@@ -6,11 +6,24 @@
 import inspect
 import re
 
+def extract_full_summary_from_signature(operation):
+    """ Extract the summary from the doccomments of the command. """
+    lines = inspect.getdoc(operation)
+    regex = re.compile(r'\s*(:param)\s+(.+)\s*:(.*)')
```
Comment (johanste) [inline src/azure/cli/commands/_introspection.py]: We should not compile a regex for each command we register. Performance.

**Comment by tjprescott** [comment]
> Command

Also, I will get getting rid of this file before I merge. I used it to diff and make sure my changes for fixing the issue with : didn't break other things. 

---
Refers to: help_dump.txt:2 in 0d7d269. [](commit_id = 0d7d26977e1a5d6f36d5aea5cdcd3034e03f5504, deletion_comment = False)

**Comment by BurtBiel** [comment]
>                       max_name_length + indent*4 + 2)

let's factor this and the version in your code out into a helper method like _get_hanging_indent.  We might as well add a comment in that method or a constant for why we do the +2 (for the colon + space)

---
Refers to: src/azure/cli/_help.py:106 in 0d7d269. [](commit_id = 0d7d26977e1a5d6f36d5aea5cdcd3034e03f5504, deletion_comment = False)

File: src/azure/cli/_help.py [hunk 2]
```diff
@@ -212,7 +213,14 @@ def __init__(self, delimiters):
...
 
     def load(self, options):
-        self.short_summary = getattr(options, 'description', None)
+        description = getattr(options, 'description', None)
+        try:
+            self.short_summary = description[:description.index('.')]
+            self.long_summary = description[description.index('.') + 1:].lstrip()
+        except (ValueError, AttributeError):
+            self.short_summary = description
+            self.long_summary = None
```
Comment (BurtBiel) [inline src/azure/cli/_help.py]: >self.long_summary = None [](start = 12, length = 24)

technically don't need this

Comment (tjprescott) [inline src/azure/cli/_help.py]: Good point. I had it the other way around originally but that caused the help tests to fail, so I changed it to conform to the tests. Unit test win. 

---
In reply to: [76083810](https://github.com/Azure/azure-cli/pull/736#discussion_r76083810) [](ancestors = 76083810)

**Comment by tjprescott** [comment]
@johanste @BurtBiel comments addressed.

File: src/azure/cli/commands/_introspection.py (hunk 1 above)
Comment (BurtBiel) [inline src/azure/cli/commands/_introspection.py]: the regex can be pulled out of the method and compiled, then using the precompiled version will be faster

---
In reply to: [75970919](https://github.com/Azure/azure-cli/pull/736#discussion_r75970919) [](ancestors = 75970919)

File: src/azure/cli/_help.py [hunk 3]
```diff
@@ -340,6 +352,9 @@ def __init__(self, _data):
         self.text = _data['text']
 
 
+def _get_hanging_indent(max_length, indent):
+    return max_length + (indent * 4) + len(FIRST_LINE_PREFIX)
+
```
Comment (BurtBiel) [inline src/azure/cli/_help.py]: nit: this could be moved next to get column indent since they are closely related

**Comment by tjprescott** [comment]
@BurtBiel comments addressed. #Closed

**Pull request merged**